### PR TITLE
Add "original" option to send original query through fallback

### PR DESCRIPTION
**Pull request by eugen-kleiner-mesh**
This implements feature https://github.com/coredns/fallback/issues/6

### CODE REVIEW

File: fallback.go [hunk 1]
```diff
@@ -36,11 +42,16 @@ func (f fallbackProxyCreator) New(trace plugin.Handler, upstream proxy.Upstream)
...
 func New(trace plugin.Handler) (f *Fallback) {
-	return &Fallback{trace: trace, rules: make(map[int]proxy.Upstream), proxy: fallbackProxyCreator{}}
+	return &Fallback{trace: trace, rules: make(map[int]rule), proxy: fallbackProxyCreator{}}
 }
 
 // ServeDNS implements the plugin.Handler interface.
 func (f Fallback) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	// If fallback has original option set for any code then copy original request to use it instead of changed
+	var originalRequest *dns.Msg
+	if f.original {
```
Comment (eugen-kleiner-mesh) [inline fallback.go:52]: This is global flag which says if we have at least one ORIGINAL option set. Needed for performance improvement.

File: fallback.go [hunk 2]
```diff
@@ -54,10 +65,13 @@ func (f Fallback) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg
 	}
 
 	if u, ok := f.rules[rules_index]; ok {
-		p := f.proxy.New(f.trace, u)
+		p := f.proxy.New(f.trace, u.proxyUpstream)
 		if p == nil {
 			return dns.RcodeServerFailure, fmt.Errorf("cannot create fallback proxy")
 		}
+		if u.original && originalRequest != nil {
```
Comment (eugen-kleiner-mesh) [inline fallback.go:72]: If ORIGINAL is set for current rule then use previously copied original request.

**Comment by yuewko** [comment]
@ekleiner , looks good.

I do think you need at least a couple of UTs to verify that the ORIGINAL option does work as intended. I suggest enhancing the `stubNextHandler` and `dummyUpstream` in `fallback_test.go` to do that.

File: setup.go [hunk 3]
```diff
@@ -24,10 +24,20 @@ func setup(c *caddy.Controller) error {
...
 	for c.Next() {
-		var rcode string
+		var (
+			original bool
+			rcode    string
+		)
 		if !c.Dispenser.Args(&rcode) {
 			return c.ArgErr()
 		}
+		if rcode == "ORIGINAL" {
```
Comment (johnbelamaric) [inline setup.go]: Why all caps?

Comment (eugen-kleiner-mesh) [reply setup.go]: Wanted to be consistent with rcode constant. But lowercase seems better here. Changed.

File: fallback_test.go [hunk 4]
```diff
@@ -52,9 +57,24 @@ func (c *testProxyCreator) New(trace plugin.Handler, upstream proxy.Upstream) pl
 		return nil
 	}
 
+	if c.handler != nil {
+		return c.handler
+	}
 	return &proxy.Proxy{Trace: trace, Upstreams: &[]proxy.Upstream{upstream}}
```
Comment (yuewko) [inline fallback_test.go]: @ekleiner , just a thought...

Maybe returning an instance of `proxy.Proxy` here is actually not very useful. How about always returning `c.handler` instead?

Looks good, btw!!

File: setup.go (hunk 3 above)
Comment (johnbelamaric) [reply setup.go]: Thanks, yes, I think lower is better, since this *isn't* an `rcode`.

File: fallback_test.go (hunk 4 above)
Comment (yuewko) [reply fallback_test.go]: .... and default to `dummyHandler`.

**Comment by yuewko** [comment]
LGTM!

**Comment by fturib** [comment]
@yuewko, @johnbelamaric  : if you are ok with last change, can you "approve" this PR ?

**Pull request merged**